torch/utils/data/graph_settings.py

```diff
@@ -36,18 +36,28 @@ def apply_sharding(datapipe: DataPipe,
                    num_of_instances: int,
                    instance_id: int,
                    sharding_group=SHARDING_PRIORITIES.DEFAULT) -> DataPipe:
+    r"""
+    Apply dynamic sharding over the ``sharding_filter`` DataPipe that has a method ``apply_sharding``.
+    RuntimeError will be raised when multiple ``sharding_filter`` are presented in the same branch.
+    """
     graph = traverse_dps(datapipe)
-    all_pipes = get_all_graph_pipes(graph)
-    already_applied_to = None
-    for pipe in all_pipes:
-        if hasattr(pipe, 'is_shardable'):
-            if pipe.is_shardable():
-                if hasattr(pipe, 'apply_sharding'):
-                    if already_applied_to is not None:
-                        raise RuntimeError('This implementation of sharding can be only applied once per instance of DataPipeline.',
-                                           'Already applied to', already_applied_to, 'while trying to apply to', pipe)
-                    pipe.apply_sharding(num_of_instances, instance_id, sharding_group=sharding_group)
-                    already_applied_to = pipe
+
+    def _helper(graph, prev_applied=None):
+        for _, (dp, sub_graph) in graph.items():
+            applied = None
+            if hasattr(dp, 'is_shardable') and dp.is_shardable():
+                if hasattr(dp, 'apply_sharding'):
+                    if prev_applied is not None:
+                        raise RuntimeError("Dynamic sharding can be only applied once per branch of DataPipe graph. "
```

Comment:
nit: let's add something like "Sharding twice is likely unintended, but if you would like to shard twice, please use filter instead"

Reply:
Sounds good. In terms of using filter, I prefer not mentioning because it probably not comes from users' intention even though double `sharding_filter` behaves like a `filter` in worker.